File: backend/app/schemas/executive.py
"""
Pydantic Schemas for Executive Cloud Operations Command Center.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

# ── Health Score & Components ──────────────────────────────────────────────────


class ComponentScore(BaseModel):
    name: str
    score: int
    weight_pct: int
    status: str  # OPTIMAL, ACCEPTABLE, RISK, CRITICAL
    details: str


class HealthScoreResponse(BaseModel):
    overall_score: int
    reliability_score: int
    security_score: int
    cost_score: int
    performance_score: int
    capacity_score: int
    governance_score: int
    incident_health: int
    risk_level: str  # HEALTHY, LOW_RISK, MODERATE_RISK, HIGH_RISK, CRITICAL
    trend: str  # IMPROVING, STABLE, WORSENING, INSUFFICIENT_DATA
    components: list[ComponentScore]
    explanation: str


# ── Executive Summary ──────────────────────────────────────────────────────────


class ExecutiveSummaryResponse(BaseModel):
    summary_text: str
    source: str  # AI-powered Summary | Local Operations Intelligence
    generated_at: datetime
    key_highlights: list[str]


# ── Key Executive Metrics ──────────────────────────────────────────────────────


class KeyExecutiveMetricsResponse(BaseModel):
    active_incidents: int
    critical_incidents: int
    unresolved_anomalies: int
    security_findings: int
    critical_security_findings: int
    current_monthly_spend: float
    projected_spend: float
    potential_savings: float
    budget_utilization_pct: float
    capacity_risk_score: int
    policy_violations: int
    pending_remediations: int
    unhealthy_services: int
    kubernetes_risk_level: str


# ── Top Priorities Queue ───────────────────────────────────────────────────────


class ExecutivePriorityItem(BaseModel):
    id: str
    priority_score: float
    priority_level: str  # P0, P1, P2, P3
    severity: str  # CRITICAL, HIGH, MEDIUM, LOW
    domain: str  # INCIDENT, SECURITY, FINOPS, CAPACITY, PERFORMANCE, GOVERNANCE, KUBERNETES
    title: str
    description: str
    affected_resource: str
    business_impact: str
    financial_impact: str
    recommended_action: str
    status: str
    created_at: datetime


class ExecutivePriorityListResponse(BaseModel):
    priorities: list[ExecutivePriorityItem]
    total: int
    p0_count: int
    p1_count: int


# ── Operational Trends ─────────────────────────────────────────────────────────


class OperationalTrendItem(BaseModel):
    metric_name: str
    domain: str
    current_period: float
    previous_period: float
    percentage_change: float | None = None
    direction: str  # UP, DOWN, NEUTRAL
    trend_status: str  # IMPROVING, STABLE, WORSENING, INSUFFICIENT_DATA
    unit: str


class OperationalTrendsResponse(BaseModel):
    trends: list[OperationalTrendItem]


# ── Cloud Provider Health ──────────────────────────────────────────────────────


class ProviderHealthItem(BaseModel):
    provider: str  # AWS, Azure, GCP, Kubernetes
    health_score: int
    monthly_spend: float
    active_incidents: int
    security_risk_level: str
    capacity_risk_score: int
    policy_violations: int
    service_count: int
    trend: str


class CloudProviderHealthResponse(BaseModel):
    providers: list[ProviderHealthItem]


# ── Service Health Map ─────────────────────────────────────────────────────────


class ServiceHealthItem(BaseModel):
    id: str
    name: str
    status: str  # HEALTHY, DEGRADED, CRITICAL, UNKNOWN
    environment: str
    provider: str
    incident_count: int
    anomaly_count: int
    monthly_cost: float
    security_findings_count: int
    capacity_risk: str
    dependencies_count: int
    last_updated: datetime


class ServiceHealthMapResponse(BaseModel):
    services: list[ServiceHealthItem]
    healthy_count: int
    degraded_count: int
    critical_count: int


# ── Cloud Risk Matrix ──────────────────────────────────────────────────────────


class RiskMatrixItem(BaseModel):
    domain: str  # Reliability, Security, FinOps, Capacity, Governance, Kubernetes, Performance
    risk_level: str  # LOW, MEDIUM, HIGH, CRITICAL
    severity: str
    trend: str
    impact_summary: str
    recommended_action: str


class CloudRiskMatrixResponse(BaseModel):
    matrix: list[RiskMatrixItem]


# ── What Changed (Delta) ───────────────────────────────────────────────────────


class ChangeItem(BaseModel):
    category: str
    metric: str
    current_value: str
    previous_value: str
    change_type: str  # INCREASE, DECREASE, NEW, RESOLVED, UNCHANGED
    significance: str  # HIGH, MEDIUM, LOW


class WhatChangedResponse(BaseModel):
    changes: list[ChangeItem]
    period_days: int = 30


# ── Executive Timeline ─────────────────────────────────────────────────────────


class ExecutiveTimelineEvent(BaseModel):
    id: str
    timestamp: datetime
    domain: str
    severity: str
    title: str
    resource: str
    status: str
    details: str | None = None


class ExecutiveTimelineResponse(BaseModel):
    events: list[ExecutiveTimelineEvent]
    total: int


# ── Executive Alerts & Recommendations ─────────────────────────────────────────


class ExecutiveAlertItem(BaseModel):
    id: str
    severity: str  # CRITICAL, HIGH, MEDIUM, LOW
    domain: str
    title: str
    message: str
    timestamp: datetime


class ExecutiveAlertsResponse(BaseModel):
    alerts: list[ExecutiveAlertItem]


class ExecutiveRecommendationItem(BaseModel):
    id: str
    domain: str
    action: str
    title: str
    impact: str
    risk_level: str
    estimated_savings: float | None = None
    suggested_owner: str
    status: str  # OPEN, IN_PROGRESS, COMPLETED


class ExecutiveRecommendationsResponse(BaseModel):
    recommendations: list[ExecutiveRecommendationItem]


# ── Executive Overview Aggregation Response ────────────────────────────────────


class ExecutiveOverviewResponse(BaseModel):
    health_score: HealthScoreResponse
    summary: ExecutiveSummaryResponse
    metrics: KeyExecutiveMetricsResponse
    top_priorities: list[ExecutivePriorityItem]
    provider_health: list[ProviderHealthItem]
    operational_trends: list[OperationalTrendItem]
    risk_matrix: list[RiskMatrixItem]
    what_changed: list[ChangeItem]
    alerts: list[ExecutiveAlertItem]
    mode_indicator: str = Field(
        default="DEMO / LOCAL MODE — Executive Intelligence Operating on Real Aggregated Platform Data",
        description="Execution mode indicator",
    )

File: frontend/src/types/executive.ts
export interface HealthComponentItem {
  name: string;
  score: number;
  weight_pct: number;
  status: string;
  details: string;
}

export interface HealthScoreResponse {
  overall_score: number;
  reliability_score: number;
  security_score: number;
  cost_score: number;
  performance_score: number;
  capacity_score: number;
  governance_score: number;
  incident_health: number;
  risk_level: 'HEALTHY' | 'LOW_RISK' | 'MODERATE_RISK' | 'HIGH_RISK' | 'CRITICAL';
  trend: 'IMPROVING' | 'STABLE' | 'WORSENING';
  components: HealthComponentItem[];
  explanation: string;
}

export interface ExecutiveSummaryResponse {
  summary_text: string;
  source: string;
  generated_at: string;
  key_highlights: string[];
}

export interface KeyExecutiveMetricsResponse {
  active_incidents: number;
  critical_incidents: number;
  unresolved_anomalies: number;
  security_findings: number;
  critical_security_findings: number;
  current_monthly_spend: number;
  projected_spend: number;
  potential_savings: number;
  budget_utilization_pct: number;
  capacity_risk_score: number;
  policy_violations: number;
  pending_remediations: number;
  unhealthy_services: number;
  kubernetes_risk_level: string;
}

export interface ExecutivePriorityItem {
  id: string;
  priority_score: number;
  priority_level: 'P0' | 'P1' | 'P2' | 'P3';
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  domain: 'INCIDENT' | 'SECURITY' | 'FINOPS' | 'CAPACITY' | 'GOVERNANCE' | 'KUBERNETES' | 'PERFORMANCE';
  title: string;
  description: string;
  affected_resource: string;
  business_impact: string;
  financial_impact: string;
  recommended_action: string;
  status: string;
  created_at: string;
}

export interface ExecutivePriorityListResponse {
  priorities: ExecutivePriorityItem[];
  total: number;
  p0_count: number;
  p1_count: number;
}

export interface OperationalTrendItem {
  metric_name: string;
  domain: string;
  current_period: number;
  previous_period: number;
  percentage_change: number;
  direction: 'UP' | 'DOWN' | 'FLAT';
  trend_status: 'IMPROVING' | 'STABLE' | 'WORSENING' | 'INSUFFICIENT_DATA';
  unit: string;
}

export interface OperationalTrendsResponse {
  trends: OperationalTrendItem[];
}

export interface ProviderHealthItem {
  provider: 'AWS' | 'Azure' | 'GCP' | 'Kubernetes' | string;
  health_score: number;
  monthly_spend: number;
  active_incidents: number;
  security_risk_level: string;
  capacity_risk_score: number;
  policy_violations: number;
  service_count: number;
  trend: string;
}

export interface CloudProviderHealthResponse {
  providers: ProviderHealthItem[];
}

export interface ServiceHealthMapItem {
  id: string;
  name: string;
  status: 'HEALTHY' | 'DEGRADED' | 'CRITICAL' | 'UNKNOWN';
  environment: string;
  provider: string;
  incident_count: number;
  anomaly_count: number;
  monthly_cost: number;
  security_findings_count: number;
  capacity_risk: string;
  dependencies_count: number;
  last_updated: string;
}

export interface ServiceHealthMapResponse {
  services: ServiceHealthMapItem[];
  healthy_count: number;
  degraded_count: number;
  critical_count: number;
}

export interface RiskMatrixItem {
  domain: string;
  risk_level: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  severity: string;
  trend: string;
  impact_summary: string;
  recommended_action: string;
}

export interface CloudRiskMatrixResponse {
  matrix: RiskMatrixItem[];
}

export interface WhatChangedItem {
  category: string;
  metric: string;
  current_value: string;
  previous_value: string;
  change_type: 'INCREASE' | 'DECREASE' | 'NEW' | 'RESOLVED' | 'STABLE';
  significance: 'HIGH' | 'MEDIUM' | 'LOW';
}

export interface WhatChangedResponse {
  changes: WhatChangedItem[];
  period_days: number;
}

export interface ExecutiveTimelineEvent {
  id: string;
  timestamp: string;
  domain: string;
  severity: string;
  title: string;
  resource: string;
  status: string;
  details: string;
}

export interface ExecutiveAlertItem {
  id: string;
  severity: string;
  domain: string;
  title: string;
  message: string;
  timestamp: string;
}

export interface ExecutiveAlertsResponse {
  alerts: ExecutiveAlertItem[];
}

export interface ExecutiveRecommendationItem {
  id: string;
  domain: string;
  action: string;
  title: string;
  impact: string;
  risk_level: string;
  estimated_savings: number;
  suggested_owner: string;
  status: string;
}

export interface ExecutiveRecommendationsResponse {
  recommendations: ExecutiveRecommendationItem[];
}

export interface ExecutiveOverviewResponse {
  health_score: HealthScoreResponse;
  summary: ExecutiveSummaryResponse;
  metrics: KeyExecutiveMetricsResponse;
  top_priorities: ExecutivePriorityItem[];
  provider_health: ProviderHealthItem[];
  operational_trends: OperationalTrendItem[];
  risk_matrix: RiskMatrixItem[];
  what_changed: WhatChangedItem[];
  alerts: ExecutiveAlertItem[];
  mode_indicator: string;
}
